File: python/1457_PseudoPalindromicPathsInABinaryTree.py
class Solution:
    def pseudoPalindromicPaths (self, root: TreeNode) -> int:
        counts = [0 for _ in range(9)]
        def helper(node, length):
            if not node:
                return 0
            
            total = 0
            counts[node.val - 1] += 1
            length += 1
            if not node.right and not node.left:
                total = pseudo(length)
            else:
                total = helper(node.left, length) + helper(node.right, length)
            length -= 1
            counts[node.val - 1] -= 1
            
            return total
        
        def pseudo(length):
            length %= 2
            for n in counts:
                if n % 2:
                    length -= 1
                    if length < 0:
                        return 0
            return 1
        
        return helper(root, 0)

File: cpp/1457_PseudoPalindromicPathsInABinaryTree.cpp
class Solution {
public:
    int pseudoPalindromicPaths (TreeNode* root) {
        unordered_map<int, int> path;
        return traverse(root, path, 0);
    }
    
    int traverse(TreeNode* root, unordered_map<int, int>& path, int length) {
        if (!root) {
            return 0;
        } 
        
        if (path.find(root->val) == path.end()) path[root->val] = 1;
        else path[root->val] += 1;
        
        int sum;
        if (!root->left && !root->right) {
            sum = palindrome(path, length + 1) ? 1 : 0;
        } else {
            sum = traverse(root->left, path, length + 1) + traverse(root->right, path, length + 1);
        }
        path[root->val] -= 1;
        return sum;
    }
    
    bool palindrome(unordered_map<int, int>& path, int length) {
        int count = length % 2;
        // cout << "PATH" << endl;
        for (auto entry : path) {
            int v = entry.second;
            // cout << "n: " << entry.first << ", count: " << v << endl;
            if (entry.second % 2 && --count < 0) {
                return false;
            }
        }
        // cout << "TRUE" << endl;
        return true;
    }
};
